### PR TITLE
Inf evaluator

### PR DESCRIPTION
**Pull request by armintoepfer**
- Remove AccumulateNoInf
 - Add InvalidEvaluatorException
 - AbstractIntegrator::LLs(Mutation) will invalide an Evaluator that returns -inf and throw an InvalidEvaluatorException
 - Polish can catch InvalidEvaluatorException and retry mutation finding with smaller list of Evaluators; in this case, Polish will warn via std::cerr.
 - Polish::ConsensusQuality(ai) skips over InvalidEvaluatorExceptions

### CODE REVIEW

File: src/AbstractIntegrator.cpp [hunk 1]
```diff
@@ -89,31 +92,64 @@ Data::State AbstractIntegrator::AddRead(std::unique_ptr<AbstractTemplate>&& tpl,
...
         }
-    };
 
-    return TransformEvaluators<double>(functor);
+        // If the mutation of interest caused a corner-case failure,
+        // release this Evaluator and report this issue via an exception.
+        if (std::isinf(ll))
+        {
+            std::string name = e.ReadName();
+            e.Release();
```
Comment (dalexander) [inline src/AbstractIntegrator.cpp]: I think "Release" is intended for client code, not for use within the AI.  It sets the state to `State::MANUALLY_RELEASED`.  I don't think that's the state we want to use here.

**Comment by dalexander** [comment]
In looking at this, I think we are hampered by the lack of API documentation on the AbstractIntegrator. So, for example, I think some of the changes here change the API contract, but it's unclear because there *is no API contract.*

File: src/Polish.cpp [hunk 2]
```diff
@@ -295,7 +318,15 @@ vector<int> ConsensusQualities(AbstractIntegrator& ai)
...
             // TODO (lhepler): this is dumb, but untestable mutations,
             //   aka insertions at ends, cause all sorts of weird issues
-            const double score = ai.LL(m) - LL;
+            double score;
+            try {
+                score = ai.LL(m) - LL;
+            } catch (const Exception::InvalidEvaluatorException& e) {
+                // If an Evaluator exception occured, report and skip! 
+                // We need to handle this!
+                std::cerr << "In Polish::ConsensusQualities(ai):" << e.what() << std::endl;
```
Comment (nlhepler) [inline src/Polish.cpp]: I think at the moment we don't have an official logging solution (or dependency injection for such) for unanimity/CC2. Until we can figure out how to do that, I'm not sure this is precisely what we should do...

Comment (dalexander) [inline src/Polish.cpp]: I was about to comment along these lines.  ConsensusCore[2] has never had "true" logging mechanism.  ConsensusCore had a trivial diagnostic logging capability here that enabled it to be turned on/off:

https://github.com/PacificBiosciences/ConsensusCore/blob/master/src/C%2B%2B/Logging/Logging.cpp


I agree that we shouldn't print to cerr right now, we should talk about the logging plan first

File: src/AbstractIntegrator.cpp [hunk 3]
```diff
@@ -89,31 +92,64 @@ Data::State AbstractIntegrator::AddRead(std::unique_ptr<AbstractTemplate>&& tpl,
...
+            throw InvalidEvaluatorException("-INF in mutation testing: " + name);
+        }
+        // Otherwise, store the ll
+        lls.emplace_back(ll);
+    }
+
+    return lls;
 }
 
 std::vector<double> AbstractIntegrator::LLs() const
```
Comment (dalexander) [inline src/AbstractIntegrator.cpp]: This changes the contract of the method.  I think we need to write some API documentation here that describes what the contracts are.  In particular, what should the "invalid" value be?  I would argue for nan, over 0...

**Comment by dalexander** [comment]
As to the substance of the commit... I'm still thinking about it

File: src/AbstractIntegrator.cpp (hunk 3 above)
Comment (nlhepler) [inline src/AbstractIntegrator.cpp]: I agree with that, both the present contracts and the nan value.

**Comment by nlhepler** [comment]
This definitely implements what we discussed: the sanest most conservative approach for how we should address this possibility.

I'm not sure we should be creating temporary std::vectors only to accumulate them afterwards -- I'm not confident they'll be elided away entirely, it may have a performance impact (presently testing mutations does no allocation that I'm aware of). We might want a private function that does both via returning the sum and supporting per-eval LLs via an output parameter std::vector<double>* and if !nullptr, clears, reserves, and fills it.

**Comment by armintoepfer** [comment]
This is mainly a POC that what we discussed works. I wasn't prepared that you rip the core of CC2 apart, because it lacks logging, contracts, and documentation. Of course I'm changing the contracts, as it is required. I'd prefer @dalexander puts this as a story into jira and split it there.

**Pull request closed without merging**